cpp/src/c_api/random_walks.cpp

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cugraph_c/algorithms.h>
+
+#include <c_api/abstract_functor.hpp>
+#include <c_api/graph.hpp>
+#include <c_api/utils.hpp>
+
+#include <cugraph/algorithms.hpp>
+#include <cugraph/detail/utility_wrappers.hpp>
+#include <cugraph/graph_functions.hpp>
+#include <cugraph/visitors/generic_cascaded_dispatch.hpp>
+
+#include <raft/handle.hpp>
+
+namespace cugraph {
+namespace c_api {
+
+struct cugraph_random_walk_result_t {
+  bool result_compressed_;
+  size_t max_path_length_;
+  cugraph_type_erased_device_array_t* paths_;
+  cugraph_type_erased_device_array_t* weights_;
+  cugraph_type_erased_device_array_t* sizes_;
+};
+
+struct node2vec_functor : public abstract_functor {
+  raft::handle_t const& handle_;
+  cugraph_graph_t* graph_;
+  cugraph_type_erased_device_array_view_t const* sources_;
+  size_t max_depth_;
+  bool compress_result_;
+  double p_;
+  double q_;
+  cugraph_random_walk_result_t* result_{};
```

Comment:
Isn't it our convention to define member variables before functions? And don't forget {} at the end of variable names.

Reply:
Added the {}.  Not sure what you mean by your question.

Reply:
What we typically have is
```
struct {
  functions(); ...
  variables; ...
};
```
but here
```
struct {
  variables;...
  functions(); ...
 };
```

Reply:
Any reason for this?

Reply:
Ah.  I interpreted your question as the opposite - hence my confusion.

What I have generally seen in our code base is that when we define a class we define it as:

class {
public:
   functions;
private:
   private_functions;
   variables;
};

But when we define a struct (essentially the same, only real difference is that everything is public) we define it as:

struct {
  variables;
  functions;
};

Admittedly, *most* of the structs that we define are to work around the nvcc problem with lambdas.  My feeling (from a design perspective) was that this functor is really a heavier weight lambda.  But you are correct in observing that it is more class-like.  It does derive from an abstract struct, hence the need for a constructor as well as an `operator()()`.

This is consistent across all of the C API functions.  We can certainly change it if we like, but if we do we should fix the entire C API (and better to do it before we add too many more calls).

Reply:
And I guess the sampling_params_t is a struct but follows the class description in my above comment.  So we have inconsistency there as well.

Reply:
Oh, yes, you're right, actually, I also do {vairables; operators();} for struct (i.e. functors). Seeing https://github.com/rapidsai/cugraph/pull/2069/files#diff-1264d5c0712616b080bb085897241ca6ad0bd8d4d132c83a75be4edf229ce988L28, I erroneously thought that we do functions first (do we consider constructors as something special, or should this better be class, or should we resolve inconsistencies in ordering?).

Reply:
@seunghwak - I'll make a note to consider this and write up some options for us to discuss.